### PR TITLE
[WIP] Add nn.ModuleDict (#4048)

### PR DESCRIPTION
**Pull request by rkaplan**
Issue #4048

This PR implements an nn.ModuleDict class. Its purpose is similar to nn.ModuleList (ensuring modules in a collection are properly registered) but it exposes a dict interface instead. See issue #4048 for discussion.

This PR is still untested as I have been unable to compile PyTorch from source on my Mac; please do not merge it yet. I am putting it out for feedback now until I find the chance to sit down and fix the compilation issues and test it myself.

This is my first contribution to PyTorch, please let me know if I should be doing anything differently. Cheers.

### CODE REVIEW

File: torch/nn/modules/container.py [hunk 1]
```diff
@@ -90,6 +90,78 @@ def forward(self, input):
...
+    def __init__(self, modules=None):
+        super(ModuleDict, self).__init__()
+        if modules is not None:
+            self.update(modules)
+
+    def __getitem__(self, key):
+        return self._modules[key]
+
+    def __setitem__(self, key, module):
+        return setattr(self, key, module)
```
Comment (SsnL) [inline torch/nn/modules/container.py]: I don't like that setting `a_module_dict['keys'] = my_module' breaks the `keys()` method.

Comment (rkaplan) [reply torch/nn/modules/container.py]: Sorry, why does this break .keys()? In your example, `__setattr__` in `nn.Module` will add `my_module` to `self._modules`: https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L426

So if I'm not mistaken calling .keys() should still behave as expected, returning `keys` : `my_module` in addition to what was already there.

File: torch/nn/modules/container.py [hunk 2]
```diff
@@ -90,6 +90,78 @@ def forward(self, input):
...
+
+    def get(self, key, default=None):
+        return self._modules.get(key, default)
+
+    def update(self, modules):
+        r"""Updates modules from a Python dict.
+        Arguments:
+            modules (dict): dict of modules to append
+        """
+        if not isinstance(modules, dict):
```
Comment (dhpollack) [inline torch/nn/modules/container.py]: I think you should allow updating from other structures.  Notably the `named_modules` function, which is part of the `nn.Module` creates a generator of (name, value) tuples.

**Comment by ezyang** [comment]
@rkaplan Are you planning to fix the tests?

**Comment by sbhaaf** [comment]
Here is mine
```
class ModuleDict(nn.Module):
    def __init__(self, data=None):
        super().__init__()
        self._module_list = nn.ModuleList()
        self._module_idcs = {}
        self._items = {}
        self.__dict__['_save_params'] = {}
        if data is not None:
            if isinstance(data, dict):
                for key,value in data.items():
                    self[key] = value
            elif all(len(item)==2 for item in list_(data)):
                for key,value in list_(data):
                    self[key] = value

    def forward(self, *inputs, **kwargs):
        raise NotImplementedError

    def __setitem__(self, key, value):
        def _has_module(arg):
            if isinstance(arg, dict): return any(_has_module(arg[k]) for k in arg)
            return isinstance(arg, nn.Module)

        if key in self._items:
            del self[key]

        if isinstance(value, dict):
            if _has_module(value): self[key] = ModuleDict(value)
            else: self._items[key] = value
        elif isinstance(value, nn.Module):
            self._items[key] = value
            self._module_idcs[key] = len(self._module_list)
            self._module_list.append(value)
        else:
            self._items[key] = value
            self.__dict__['_save_params'][key] = value

    def __getitem__(self, key):
        if key in self._items:
            return self._items[key]
        raise KeyError('ModuleDict as no attribute `%s`' % key)

    def __delitem__(self, key):
        if isinstance(self[key], nn.Module):
            self._module_list[self._module_idcs[key]] = None
            del self._module_idcs[key]
        else:
            del self.__dict__['_save_params'][key]
        del self._items[key]

    def __repr__(self, root=True):
        return ('ModuleDict({%s})' if root else '{%s}') % ', '.join(
            ["'%s': %s" % (k, v.__repr__(False) if isinstance(v, ModuleDict) else v if not isinstance(v, nn.Module) \
                else v.__class__.__name__ + '()') for k,v in self.items()]
        )

    def __iter__(self):
        return iter(self._items)

    def items(self):
        return self._items.items()

    def keys(self):
        return self._items.keys()

    def init(self):
        return init_([mod for mod in self._module_list])
```

**Comment by karandwivedi42** [comment]
@rkaplan I can make this ready if you are busy.

**Comment by ezyang** [comment]
@karandwivedi42 Go ahead :)

**Comment by vishwakftw** [comment]
This is redundant I suppose, after #8463

**Comment by SsnL** [comment]
closing in favor of #8463

**Pull request closed without merging**